src/lib/db.ts

```diff
@@ -607,8 +607,21 @@ export class DB {
         .then(([documentSnapshot, metadataSnapshot]) => {
           const document: DBDocument|null = documentSnapshot.val();
           const metadata: DBDocumentMetadata|null = metadataSnapshot.val();
-          if (!document || !metadata) {
-            throw new Error("Unable to open document");
+          if (!metadata) {
+            // if we have no metadata, there's nothing we can do
+            const msg = `Error retrieving metadata for ` +
+                        `document '${documentKey}' of type '${type}' for user '${userId}'`;
+            throw new Error(msg);
+          }
+          if (!document) {
+            // If we have metadata but no document content, we can return a valid empty document.
+            // This has been seen to occur in the wild, presumably as a result of a prior bug.
+            const msg = "Warning: Reconstituting empty contents for " +
+                        `document '${documentKey}' of type '${type}' for user '${userId}'`;
+            console.warn(msg);
+            return DocumentModel.create({
+                                  type, title, properties, groupId, visibility, uid: userId, originDoc,
+                                  key: documentKey, createdAt: metadata.createdAt, content: {}, changeCount: 0 });
```

Comment:
Will this new document get saved correctly back to Firebase?
It looks like in most places after openDocument is called the resulting document is monitored, so I guess that means if it is changed then it will be saved.
I wasn't able to see where this newly created document is added to the Document store, but I'd guess that is the responsibility of something higher up in the call chain which I didn't find. 
Or maybe it isn't necessary to add all documents to the document store?

Reply:
My first answer is that I don't think it matters for the purpose of this PR. Whatever the code path is in which this function gets called it will now return an empty document where it used to throw an exception. That seems like an improvement regardless of whether the document subsequently gets saved or monitored correctly. But that's not a very satisfying answer, so let me try to do better. 😉 The callers of `DB.openDocument` are `createDocumentModelFromProblemMetadata`, `createDocumentModelFromOtherDocument`, `createDocumentModelFromOtherPublication`, and `createDocumentFromPublication`, each of which is only called by the DB listeners to create the `DocumentModel`. The first two callers explicitly set up monitoring for problem and "other" (personal documents and learning logs) documents, while the latter two callers don't set up monitoring because publications do not require monitoring. Currently, the DB listener clients of each of these functions is then responsible for adding the returned document to the `documents` store. Now that you've asked the question, however, it seems appropriate to centralize that responsibility in the `openDocument` call itself.